File: Minesweeper.py
from random import randint
class Minesweeper:
    
    nx=0
    ny=0
    nmines=0
    grid=[[]]
    selected=[[]]
    flagged=[[]]
    
    def __init__(self, nx=10, ny=10, nmines=10):
        
        self.nx=nx
        self.ny=ny
        self.nmines=nmines
        
        self.grid=[[0 for j in range(ny)] for i in range(nx)]
        self.selected=[[False for j in range(ny)] for i in range(nx)]
        self.flagged=[[False for j in range(ny)] for i in range(nx)]
      
        for i in range(nmines):
            fl=0
            while (fl==0):
                x=randint(0, nx-1)
                y=randint(0, ny-1)
                if (self.grid[x][y]==0):
                    fl=1
                    self.grid[x][y]=-1
        
        # here put in the rest of the numbers
        for i in range(nx):
            for j in range(ny):
                if (self.grid[i][j]==-1):
                    continue
                x1=max(i-1,0)
                x2=min(i+1,nx-1)
                y1=max(j-1,0)
                y2=min(j+1,ny-1)
                for x0 in range(x1,x2+1):
                    for y0 in range(y1,y2+1):
                        if (self.grid[x0][y0]==-1):
                            self.grid[i][j]=self.grid[i][j]+1
                
    def select(self, x, y):
        self.selected[x][y]=True
        self.flagged[x][y]=False
        if (self.grid[x][y]==0):
            x1=max(x-1,0)
            x2=min(x+1,self.nx-1)
            y1=max(y-1,0)
            y2=min(y+1,self.ny-1)
            print(x1,x2,y1,y2)
            for i in range(x1,x2+1):
                for j in range(y1,y2+1):
                    if (not self.selected[i][j]):
                        z=self.select(i,j)
                        if (z==-1): # something went wrong
                            exit
        
        return self.grid[x][y]
    
    def textoutput(self):
        
        if (self.nx>9):
            # print tens line
            s="  "
            for i in range(self.nx):
                if (i<10):
                    s=s+"  "
                else:
                    s=s+" %1d" % (i/10)
            print(s)
            # print ones line
            s="  "
            for i in range(self.nx):
                s=s+" %1d" % (i%10)
            print(s)
            
            for j in range(self.ny):
                s="%2d" % j
                for i in range(self.nx):
                    if (self.flagged[i][j]):
                        s=s+" !"
                        continue
                    if (not self.selected[i][j]):
                        s=s+" ?"
                        continue
                    if (self.grid[i][j]==0):
                        s=s+" ."
                        continue
                    s=s+" %1d" % self.grid[i][j]
                print(s)
    
    def playerwon(self):
        n=0
        for i in range(self.nx):
            for j in range(self.ny):
                if (not self.selected[i][j]):
                    n=n+1
                    if (n>self.nmines):
                        return False
                if ((self.selected[i][j]) and (self.grid[i][j]==-1)):
                    return False
        
        if (n==self.nmines):
            return True
        else:
            return False
        
a=Minesweeper(15,15,20)
won=False
while (not won):
    a.textoutput()
    xstr=input("Enter x of space to check:")
    x=int(xstr)
    ystr=input("Enter y of space to check:")
    y=int(ystr)
    z=a.select(x, y)
    if (z==-1):
        print("BOOM! You hit a mine!")
        exit
    won=a.playerwon()
    
print("You got all the mines!")
    

File: Minesweeper.java
//import java.lang.*;
import java.util.Random;
import java.util.Scanner;

public class Minesweeper {
	int nx;
	int ny;
	int nmines;
	int grid[][];
	boolean selected[][];

	Minesweeper(int nx0, int ny0, int nm0) throws IndexOutOfBoundsException {
		int i, j, x0, y0;
		Random rnd=new Random();
		if (nx0>0) {
			nx=nx0;
		} else {
			throw new IndexOutOfBoundsException();
		}
		if (ny0>0) {
			ny=ny0;
		} else {
			throw new IndexOutOfBoundsException();
		}
		if ((nm0>0) && (nm0<nx*ny)) {
			nmines=nm0;
		} else {
			throw new IndexOutOfBoundsException();
		}

		grid=new int[nx][ny];
		selected=new boolean[nx][ny];

		// First put in the mines... each one is marked by a -1
		for (i=0; i<nmines; i++) {
			j=0;
			while (j==0) {
				x0=(int) (nx*rnd.nextDouble());
				y0=(int) (ny*rnd.nextDouble());
				if (grid[x0][y0]==0) {
					grid[x0][y0]=-1;
					j=1;
				}
			}
		}

		// Now put in the mine numbers 
		for (i=0; i<nx; i++) {
			for (j=0; j<ny; j++) {
				if (grid[i][j]==-1) continue;
				int x1=(i==0)?0:i-1;
				int x2=(i==nx-1)?nx-1:i+1;
				int y1=(j==0)?0:j-1;
				int y2=(j==ny-1)?ny-1:j+1;
				for (x0=x1; x0<=x2; x0++) 
					for (y0=y1; y0<=y2; y0++)
						if (grid[x0][y0]==-1) grid[i][j]++;
			}
		}
	}

	public int select(int x0, int y0) throws IndexOutOfBoundsException {
		// returns -1 if the player blew up, else number of space
		// Also handles 'spreading' if player selects a zero space
		int i,j,k,x1,y1,x2,y2;

		if ((x0<0) || (x0>nx-1) || (y0<0) || (y0>ny-1))
			throw new IndexOutOfBoundsException();

		selected[x0][y0]=true;
		if (grid[x0][y0]==0) {
			x1=(x0==0)?0:x0-1;
			x2=(x0==nx-1)?nx-1:x0+1;
			y1=(y0==0)?0:y0-1;
			y2=(y0==ny-1)?ny-1:y0+1;
			for (i=x1; i<=x2; i++)
				for (j=y1; j<=y2; j++)
					if (!selected[i][j]) {
						k=this.select(i,j);
						if (k==-1) throw new IndexOutOfBoundsException();
					}
		}
		return grid[x0][y0];
	}

	public boolean playerwon() {
		int i,j;
		int nspacesleft=0;
		for (i=0; i<nx; i++)
			for (j=0; j<ny; j++)
				if (!selected[i][j]) {
					if (++nspacesleft > nmines) return false;
				}
		// double-check
		if (nspacesleft==nmines) return true;
		return false;
	}

	public void textdisplay() {
		int i,j;
		if (nx>9) {
			System.out.print("  ");
			for (i=0; i<nx; i++) {
				if (i<10) {
					System.out.print("  ");
				} else {
					j=i/10;
					System.out.print(" ");
					System.out.print(Integer.toString(j));
				}
			}
			System.out.println();
		}
		System.out.print("  ");
		for (i=0; i<nx; i++) {
			j=i%10;
			System.out.print(" ");
			System.out.print(Integer.toString(j));
		}
		System.out.println();
		System.out.println();

		for (j=0; j<ny; j++) {
			if (j<10) {
				System.out.print(" "+Integer.toString(j));
			} else {
				System.out.print(Integer.toString(j));
			}
			for (i=0; i<nx; i++) {
				System.out.print(" ");
				if (selected[i][j]) {
					System.out.print(Integer.toString(grid[i][j]));
				} else {
					System.out.print("?");
				}
			}
			System.out.println();
		}			
	}

	public static void main(String args[]) {
		Minesweeper a;
		boolean won=false;
		int i,j,k;
		Scanner reader=new Scanner(System.in);

		a=new Minesweeper(15,15,12);
		
		do {
			a.textdisplay();
			System.out.println("Enter coordinates to check:");
			i=reader.nextInt();
			j=reader.nextInt();
			if ((i>=0) && (i<a.nx) && (j>=0) && (j<=a.ny)) {
				k=a.select(i,j);
				if (k==-1) {
					System.out.println("BOOM! You lose!");
					reader.close();
					System.exit(0);
				}
			}
			won=a.playerwon();
		} while (!won);
		System.out.println("You avoided all the mines!");
		a.textdisplay();
		reader.close();
	}

}